pkg/webhook/v1alpha2/applicationconfiguration/validating_handler.go

```diff
@@ -23,25 +21,40 @@ import (
 )
 
 const (
-	reasonFmtWorkloadNameNotEmpty = "Versioning-enabled component's workload name MUST NOT be assigned. Expect workload name %q to be empty."
+	errFmtWorkloadNameNotEmpty = "versioning-enabled component's workload name MUST NOT be assigned, expect workload name %q to be empty."
 
-	errFmtCheckWorkloadName = "Error occurs when checking workload name. %q"
+	errFmtRevisionName = "componentName %q and revisionName %q are mutually exclusive, you can only specify one of them"
 
-	errFmtUnmarshalWorkload = "Error occurs when unmarshal workload of component %q error: %q"
+	errFmtUnappliableTrait = "the trait %q cannot apply to workload %q of component %q (appliable: %q)"
 
 	// WorkloadNamePath indicates field path of workload name
 	WorkloadNamePath = "metadata.name"
 )
 
 var appConfigResource = v1alpha2.SchemeGroupVersion.WithResource("applicationconfigurations")
 
+// AppConfigValidator provides functions to validate ApplicationConfiguration
+type AppConfigValidator interface {
+	Validate(context.Context, client.Client, discoverymapper.DiscoveryMapper, ValidatingAppConfig) []error
```

Comment:
I see most cases only `ValidatingAppConfig ` is used, why don't make it more clean like:
```suggestion
	Validate(context.Context, ValidatingAppConfig) []error
```

Reply:
Maybe ValidateFunc needs to use client&dm in its own logic in the future, even though current ValidateFuncs don't need because prepared data is sufficient for them.